sklearn/utils/_metadata_requests.py

```diff
@@ -0,0 +1,1129 @@
+
+        If any metadata is passed which is marked as WARN, a warning is raised.
+
+        Parameters
+        ----------
+        params : dict
+            The metadata passed to a method.
+        """
+        params = {} if params is None else params
+        warn_params = {
+            prop
+            for prop, alias in self._requests.items()
+            if alias == RequestType.WARN and prop in params
+        }
+        for param in warn_params:
+            warn(
+                f"Support for {param} has recently been added to this class. "
+                "To maintain backward compatibility, it is ignored now. "
+                "You can set the request value to RequestType.UNREQUESTED "
+                "to silence this warning, or to RequestType.REQUESTED to "
+                "consume and use the metadata."
+            )
+
+    def _route_params(self, params=None):
+        """Return the input parameters requested by the method.
+
+        The output of this method can be used directly as the input to the
+        corresponding method as extra props.
+
+        Parameters
+        ----------
+        params : dict
+            A dictionary of provided metadata.
+
+        Returns
+        -------
+        params : Bunch
+            A :class:`~utils.Bunch` of {prop: value} which can be given to the
+            corresponding method.
+        """
+        self._check_warnings(params=params)
+        params = {} if params is None else params
+        args = {arg: value for arg, value in params.items() if value is not None}
+        res = Bunch()
+        for prop, alias in self._requests.items():
+            try:
+                alias = RequestType(alias)
+            except ValueError:
+                # we leave alias as is (str) if it's not a RequestType
+                pass
+
+            if alias == RequestType.UNREQUESTED or alias == RequestType.WARN:
+                continue
+            elif alias == RequestType.REQUESTED and prop in args:
+                res[prop] = args[prop]
+            elif alias == RequestType.ERROR_IF_PASSED and prop in args:
+                raise ValueError(
+                    f"{prop} is passed but is not explicitly set as "
+                    f"requested or not for {self.owner}.{self.method}"
+                )
+            elif alias in args:
+                res[prop] = args[alias]
+        return res
+
+    def _serialize(self):
+        """Serialize the object.
+
+        Returns
+        -------
+        obj : dict
+            A serialized version of the instance in the form of a dictionary.
+        """
+        result = dict()
+        # Then parameters with string aliases
+        result.update(
+            {
+                prop: alias
+                for prop, alias in self._requests.items()
+                if alias not in RequestType
+            }
+        )
+        # And at last the parameters with RequestType routing info
```

Comment:
why do these need to be handled separately? Won't a single dict comprehension do the trick?

Reply:
apparently dict comprehension is not as comprehensive as list comprehension. It doesn't support the same complex syntax as list does.

Reply:
I don't understand. Are these 17 lines not equivalent to:
```
result = {
    prop: RequestType(alias) if RequestType.is_valid(alias) else alias
    for prop, alias in self._requests.items()
}
```